AMDGPU/GlobalISel: Temporal divergence lowering i1

petar-avramovic opened the pull request:
Use of i1 outside of the cycle, both uniform and divergent,
is lane mask(in sgpr) that contains i1 at iteration that lane
exited the cycle.
Create phi that merges lane mask across all iterations.

petar-avramovic commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/124299?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#124299** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/124299?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/124299?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#124298** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/124298?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#124297** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/124297?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

petar-avramovic commented:
Insert point of merging phi is changed to after Inst, not in the exiting block.

nhaehnle commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp:
```diff
@@ -219,6 +220,54 @@ bool DivergenceLoweringHelper::lowerTemporalDivergence() {
+    Register MergedMask = MRI->createVirtualRegister(BoolS1);
+    Register PrevIterMask = MRI->createVirtualRegister(BoolS1);
+
+    MachineBasicBlock *CycleHeaderMBB = Cycle->getHeader();
+    SmallVector<MachineBasicBlock *, 1> ExitingBlocks;
+    Cycle->getExitingBlocks(ExitingBlocks);
+    assert(ExitingBlocks.size() == 1);
+    MachineBasicBlock *CycleExitingMBB = ExitingBlocks[0];
+
+    B.setInsertPt(*CycleHeaderMBB, CycleHeaderMBB->begin());
+    auto CrossIterPHI = B.buildInstr(AMDGPU::PHI).addDef(PrevIterMask);
+
+    // We only care about cycle iterration path - merge Reg with previous
+    // iteration. For other incomings use implicit def.
+    // Predecessors should be CyclePredecessor and CycleExitingMBB.
+    // In older versions of irreducible control flow lowering there could be
+    // cases with more predecessors. To keep this lowering as generic as
+    // possible also handle those cases.
+    for (auto MBB : CycleHeaderMBB->predecessors()) {
+      if (MBB == CycleExitingMBB) {
+        CrossIterPHI.addReg(MergedMask);
+      } else {
+        B.setInsertPt(*MBB, MBB->getFirstTerminator());
+        auto ImplDef = B.buildInstr(AMDGPU::IMPLICIT_DEF, {BoolS1}, {});
+        CrossIterPHI.addReg(ImplDef.getReg(0));
+      }
+      CrossIterPHI.addMBB(MBB);
+    }
```
This still feels vaguely off to me.

First of all, can you please add a `Cycle->isReducible()` assert here just in case, since you're relying on there only being one header? (Actually, that assert should probably be in `GenericCycle::getHeader`.)

Second, I'm worried about that case distinction between different kinds of predecessors. It really doesn't feel properly generic to me. Latches and exiting blocks aren't necessarily the same thing. (The current structurizer tends to make it that way, but even with the structurizer we probably don't have a guarantee, let alone with a future different control flow lowering.)

Let's think through some cases:

* If the predecessor is outside the cycle, it should be an `IMPLICIT_DEF`
* If it is inside the cycle, i.e. a latch, there are two cases:
  * If the predecessor is dominated by `Inst`, then clearly use `MergedMask`
  * Otherwise, it's actually not clear: could it be possible that control flow leaves the dominance region of `Inst` and goes around the loop again?

On that last point, you could perhaps make a case that because the loop is temporally divergent, having the CFG in wave CFG form requires a certain structure. But I'm not convinced. In fact, consider:
```mermaid
flowchart TD
    Header --> B
    Header --> Inst
    Inst --> B
    B --> Header
    Inst --> C
    C --> Header
    C --> Exit
```
Let's say the branches of Header and C are uniform; only the branch of the basic block containing Inst is divergent.

Then we have temporal divergence in Inst, and need to handle the `i1`. But the CFG is already *reconvergent*: the divergent branch is post-dominated by one of its immediate successors. This means it is possible to insert EXEC masking code for this CFG correctly without structuring it. Therefore, the code here should really handle this case correctly.

I think this calls for using MachineSSAUpdater for a robust solution:

* Initialize the updater with IMPLICIT_DEF in every predecessor of the header that is outside the cycle.
* Insert the MergedMask into the updater for the `Inst->getParent()`
* Then build the lane mask merging, querying the updater for the previous mask (i.e., let the updater create that register) 
* The MergedMask can still be used by UseInst due to dominance

Except now there's an additional issue with caching the MergedMask that doesn't exist for the non-i1 case: Different `UseInst`s could appear with respect to different `Cycle`s. The non-i1 case doesn't actually care about the cycle structure, but the code here does.

In particular, with my suggestions, it will use the cycle to determine how to initialize the updater.

We can still get away with a single `MergedMask` per `Inst`:

* The lazy way to do it would be to not initialize the updater with IMPLICIT_DEFs and just let it figure them out by itself. This would generate correct code.
  * However, this has a compile-time cost and could end up being too pessimistic in terms of the generated code.
  * The Inst and all its uses could be contained in a cycle. This naive way would lead to a phi node at the header of this cycle, and therefore needlessly extending the live range of the value to the entire outer cycle. In this example, the values we create only need to live from InnerHeader to UseInst, but the naive use of the updater will extend them from OuterHeader all the way through OuterLatch:
```mermaid
flowchart TD
    Start --> OuterHeader --> InnerHeader --> Inst --> InnerHeader
    Inst --> UseInst --> OuterLatch --> OuterHeader
    OuterLatch --> Exit
```
* The smarter way is to initialize the updater in the way I described, using the *largest* relevant cycle.

So instead of merely caching the `MergedMask`, we should collect all `UseInst`s and determine the largest relevant cycle (as returned by the uniformity analysis).

nhaehnle reviewed (commented):
Thank you for moving the insert point.

An analogous comment applies here like in the non-i1 case: it would be good to cache the `MergedMask` register for re-use when the same `Inst` occurs with multiple `UseInst`s.

nhaehnle commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp:
```diff
@@ -219,6 +220,54 @@ bool DivergenceLoweringHelper::lowerTemporalDivergence() {
+      CrossIterPHI.addMBB(MBB);
+    }
+
+    MachineBasicBlock *MBB = Inst->getParent();
+    buildMergeLaneMasks(*MBB, MBB->getFirstTerminator(), {}, MergedMask,
+                        PrevIterMask, Reg);
```
In this case, it would be better to move the lane merging to directly after `Inst`.

The register pressure calculus is exactly opposite from the non-i1 case:

* In the non-i1 case, shifting the COPY down is good because it reduces the live range of the VGPR while potentially making the live range of the SGPR longer. VGPRs are more expensive than SGPRs, so this is a good trade-off.
* In the i1 case, we'll have a live range for the merged mask extending across the entire cycle anyway. By moving the lane merging closer to Inst, we leave the live range of the merged mask unchanged, but we (most likely) reduce the live range of the i1 value produced by Inst.

petar-avramovic committed:
AMDGPU/GlobalISel: Temporal divergence lowering i1

Use of i1 outside of the cycle, both uniform and divergent,
is lane mask(in sgpr) that contains i1 at iteration that lane
exited the cycle.
Create phi that merges lane mask across all iterations.

nhaehnle commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp:
```diff
@@ -228,6 +229,66 @@ bool DivergenceLoweringHelper::lowerTemporalDivergence() {
+    Register Reg = LRCIter.first;
+    const MachineCycle *Cycle = LRCIter.second.first;
+
+    if (MRI->getType(Reg) != LLT::scalar(1))
+      continue;
```
This check is now redundant

nhaehnle commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp:
```diff
@@ -228,6 +229,66 @@ bool DivergenceLoweringHelper::lowerTemporalDivergence() {
+    buildMergeLaneMasks(*MBB, MBB->getFirstTerminator(), {}, MergedMask,
+                        SSAUpdater.GetValueInMiddleOfBlock(MBB), Reg);
+
+    LRCCache[Reg].second = MergedMask;
```
Should be able to just keep LRCIter/Entry as a reference and update via that instead of repeating the cache lookup.

nhaehnle commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp:
```diff
@@ -228,6 +229,66 @@ bool DivergenceLoweringHelper::lowerTemporalDivergence() {
+    LRCCache[Reg] = {LRC, {}};
+  }
+
+  for (auto LRCIter : LRCCache) {
```
Naming: this isn't an iterator, a more accurate generic name would be just "Entry" (or LRCCacheEntry, but that's long)

nhaehnle commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp:
```diff
@@ -228,6 +229,66 @@ bool DivergenceLoweringHelper::lowerTemporalDivergence() {
+    if (MRI->getType(Reg) != LLT::scalar(1))
+      continue;
+
+    replaceUsesOfRegInInstWith(Reg, UseInst, LRCCache[Reg].second);
```
Can use .lookup instead of operator[] for consistency with above.

nhaehnle commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelDivergenceLowering.cpp:
```diff
@@ -228,6 +229,66 @@ bool DivergenceLoweringHelper::lowerTemporalDivergence() {
+    if (MRI->getType(Reg) != LLT::scalar(1))
+      continue;
+
+    const MachineCycle *CachedLRC = LRCCache.lookup(Reg).first;
+    if (CachedLRC) {
+      LRC = CachedLRC->contains(LRC) ? CachedLRC : LRC;
+      assert(LRC->contains(CachedLRC));
+    }
+
+    LRCCache[Reg] = {LRC, {}};
```
This is actually a great use case for try_emplace to do the cache lookup only once:
```suggestion
    const MachineCycle *&CachedLRC = LRCCache.try_emplace(Reg);
    if (!CachedLRC || !CachedLRC->contains(LRC))
      CachedLRC = LRC;
```

nhaehnle reviewed (commented):
Thanks, this now looks good to me in terms of the overall flow. I have a bunch of nitpickier, mostly style-related comments.

petar-avramovic commented:
### Merge activity

* **Mar 12, 7:13 AM EDT**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/124299).
